### ISSUE TITLE
v12.0.1 error: not implemented: regenerator: complex pattern in catch clause

### ISSUE THREAD

**Issue by michaelmorgantalabat** [vercel/next.js#30395]
### What version of Next.js are you using?

12.0.1

### What version of Node.js are you using?

14.15.0

### What browser are you using?

chrome

### What operating system are you using?

mac

### How are you deploying your application?

NODE_OPTIONS='--inspect' next dev

### Describe the Bug

I got this error `thread '<unnamed>' panicked at 'not implemented: regenerator: complex pattern in catch clause', /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/swc_ecma_transforms_compat-0.47.1/src/es2015/regenerator/case.rs:1232:30` 
after upgrade from v11 to v12

### Expected Behavior

it should work fine like v11 !

### To Reproduce

 try {
 // my code
 } catch ({ error }: any) {
   //   handling error
 }
 
 
 after I remove the try-catch statement everything works fine !

**Comment by timneutkens** [comment]
@michaelmorgantalabat it'd be super helpful if you could provide the actual code you're running given that a normal try/catch would not affect regenerator, it would have be in an async function and such. It's likely a bug with a combination of factors in the code hence why the input is really important in order to fix this.

**Comment by timneutkens** [comment]
To be clear this is related to SWC and has nothing to do with `esmExternals`, so please don't disable that 🙏

**Comment by timneutkens** [comment]
@sveggiani it'd be very helpful to get the input code from you as well 👍 I forgot to mention we don't need a full reproduction, just the single file that fails 👍

**Comment by luisorbaiceta** [comment]
It's happening in the vercel/commerce project, in the auth components @timneutkens

**Comment by luisorbaiceta** [comment]
Maybe this helps @swc-project/swc#2248 can't check it right now

**Comment by luisorbaiceta** [comment]
This might help. Can't test it right now though. 

swc-project/swc#2248

**Comment by timneutkens** [comment]
@luisorbaiceta could you provide a reproduction? I'm not able to reproduce this issue when compiling https://github.com/vercel/commerce with auth enabled.

**Comment by luisorbaiceta** [comment]
I will grant you access to the private repo. I just took the commerce template and was messing around with it. I've a devcontainer so you can reproduce it accurately @timneutkens

**Comment by cherniavskii** [comment]
[next@12.0.2](https://github.com/vercel/next.js/releases/tag/v12.0.2) fixed the issue for me!

**Comment by michaelmorgantalabat** [comment]
same issue with next@12.0.2
![image](https://user-images.githubusercontent.com/52699466/139717675-61629a4a-76b7-4dfc-86bc-4f9550b2ac37.png)

**Comment by sveggiani** [comment]
Sorry @timneutkens , we reverted the upgrade and is it not quick right now to reproduce again. I'll try to find time today to submit the code causing the issue.

**Comment by luisorbaiceta** [comment]
Hey guys! after a little bit of research (I had to dig a bit into swc), I found that the bug is caused by object destructuring the error recieved in the catch clause of a component.

So instead of doing this:

js```
...
} catch ({errors}) {
    setMessage(errors[0].message);
    setLoading(false);
}
```

Just avoid destructuring. Like this:

js```
...
} catch (error) {
    setMessage(error.errors[0].message);
    setLoading(false);
}
```

**Comment by luisorbaiceta** [comment]
Looks like it's fixed swc-project/swc#2629

**Comment by luisorbaiceta** [comment]
I confirm that this is solved in version 12.0.3

**Comment by timneutkens** [comment]
This has indeed landed in 12.0.3 👍 thanks for confirming and the reproductions provided!

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
